sb/robot/sim_robot.py

```diff
@@ -36,21 +41,157 @@ def power(self, value):
             self._power = value
 
 
-class Motor:
-    """Represents a motor board."""
-    # This is named `Motor` instead of `MotorBoard` for consistency with pyenv
+BRAKE = 0  # 0 so setting the motors to 0 has exactly the same affect as setting the motors to BRAKE
+COAST = "coast"
+
+
+class MotorBoard(object):
+    VOLTAGE_SCALE = 1
 
     def __init__(self, robot):
-        self._robot = robot
-        self.serialnum = "SIM_MBv4"
+        self.robot = robot
+        self._motors = [0,0]
+
+    def __str__(self):
+        return "MotorBoard"
+
+    def _check_voltage(self,new_voltage):
+        if new_voltage != COAST and (new_voltage > 1 or new_voltage < -1):
+            raise ValueError(
+                'Incorrect voltage value, valid values: between -1 and 1, robot.COAST, or robot.BRAKE')
+
+    @property
+    def m0(self):
+        return self._motors[0]
+
+    @m0.setter
+    def m0(self, new_voltage):
+        self._check_voltage(new_voltage)
+        self._motors[0] = new_voltage
+
+    @property
+    def m1(self):
+        return self._motors[1]
+
+    @m1.setter
+    def m1(self, new_voltage):
+        self._check_voltage(new_voltage)
+        self._motors[1] = new_voltage
+
+    __repr__ = __str__
 
-        self.m0 = MotorChannel(robot)
-        self.m1 = MotorChannel(robot)
 
-    def __repr__(self):
-        return "Motor( serialnum = \"{0}\" ) (Simulated Motor Board v4)" \
-               .format(self.serialnum)
+class ServoBoard(object):
+    def __init__(self, robot):
+        self.robot = robot
+
+    def __str__(self):
+        return "ServoBoard"
+
+    __repr__ = __str__
+
+    ULTRASOUND_ANGLES = {
+        (6, 7): ('ahead', 0),
+        (8, 9): ('right', math.pi / 2),
+        (10, 11): ('left', -math.pi / 2),
+    }
+
+    def read_ultrasound(self, trigger_pin, echo_pin):
+        pin_pair = (trigger_pin, echo_pin)
+
+        try:
+            _, angle_offset = self.ULTRASOUND_ANGLES[pin_pair]
+        except KeyError:
+            print("There's no ultrasound module on those pins. Try:")
+            for (
+                    (trigger_pin, echo_pin),
+                    (direction, _),
+            ) in self.ULTRASOUND_ANGLES.items():
+                print("Pins {} and {} for the sensor pointing {}".format(
+                    trigger_pin,
+                    echo_pin,
+                    direction,
+                ))
+            return 0.0
```

Comment:
I think rather than returning here, we should re-raise the error. It'll help the students check what pin theyre sending it off

Reply:
This is complicated -- the real API has no feedback at all about this sort of failure mode.

Much as I see the benefit of reporting the failure here, I _really_ don't like the idea that we're training the competitors that they'll get an error if they use the wrong pins. Doing so leads to a high probability of confusing silent failures on the real robot (which is _much_ worse than a confusing error in the simulator).